docs/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to.md

```diff
@@ -0,0 +1,500 @@
+                reader.ValueSequence.ToArray() :
+                reader.ValueSpan;
+    // ...
+  }
+}
+```
+
+### Read null values
+
+Code that uses the `Newtonsoft.Json` reader can compare a value token type of String to null, as in the following example:
+
+```csharp
+public static string ReadAsString(this JsonTextReader reader)
+{
+    reader.Read();
+
+    if (reader.TokenType != JsonToken.String)
+    {
+        throw new InvalidDataException();
+    }
+
+    return reader.Value?.ToString();
+}
+```
+
+This code doesn't work correctly in <xref:System.Text.Json>, which considers the `null` literal to be its own token type. When using `Utf8JsonReader`, check for the `JsonTokenType.Null` token type, as shown in the following example:
+
+```csharp
+public static string ReadAsString(this ref Utf8JsonReader reader)
+{
+    reader.Read();
+
+    if (reader.TokenType == JsonTokenType.Null)
+    {
+        return null;
+    }
+
+    if (reader.TokenType != JsonTokenType.String)
+    {
+        throw new InvalidDataException();
+    }
+
+    return reader.GetString();
+}
+```
+
+### Multi-targeting
+
+If you need to continue to use `Newtonsoft.Json` for certain target frameworks, you can multi-target and have two implementations. However, this is not trivial and would require some `#ifdefs` and source duplication. One way to share as much code as possible is to create a `ref struct` wrapper around `Utf8JsonReader` and `Newtonsoft.Json` `JsonTextReader`. This wrapper would unify the public surface area while isolating the behavioral differences. This lets you isolate the changes mainly to the construction of the type, along with passing the new type around by reference. This is the pattern that the [.NET Core installer](https://github.com/dotnet/runtime/tree/master/src/installer) follows:
+
+* [UnifiedJsonReader.JsonTextReader.cs](https://github.com/dotnet/runtime/blob/35fe82c2e90fa345d0b6ad243c14da193fb123fd/src/installer/managed/Microsoft.Extensions.DependencyModel/UnifiedJsonReader.JsonTextReader.cs)
+* [UnifiedJsonReader.Utf8JsonReader.cs](https://github.com/dotnet/runtime/blob/35fe82c2e90fa345d0b6ad243c14da193fb123fd/src/installer/managed/Microsoft.Extensions.DependencyModel/UnifiedJsonReader.Utf8JsonReader.cs)
+
+## Utf8JsonWriter
+
+<xref:System.Text.Json.Utf8JsonWriter?displayProperty=fullName> is a high-performance way to write UTF-8 encoded JSON text from common .NET types like `String`, `Int32`, and `DateTime`. The writer is a low-level type that can be used to build custom serializers.
+
+The following sections explain recommended programming patterns for using `Utf8JsonWriter`.
+
+### Write with UTF-8 text
+
+To achieve the best possible performance while using the `Utf8JsonWriter`, write JSON payloads already encoded as UTF-8 text rather than as UTF-16 strings. For example, if you're writing string literals, consider caching them as static byte arrays, and write those instead.
+
+### WriteRawValue
```

Comment:
We should add a section on [StringEscapeHandling](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_StringEscapeHandling.htm) and how that compares to our default.

`Utf8JsonWriter`, by default, is equivalent to the caller using both `EscapeNonAscii` and `EscapeHtml` (at the same time). This is done for defense-in-depth purposes (secure by default).
For other behaviors, create your own `JavaScriptEncoder` and set the `Encoder` property in `JsonWriterOptions`.
https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonwriteroptions.encoder?view=netcore-3.1#System_Text_Json_JsonWriterOptions_Encoder

Also, it uses capital hex digits when writing characters out escaped (whereas `Newtonsoft.Json` uses lower case). This shouldn't affect the semantic value of the strings though.

```C#
var textWriter = new StringWriter();
using (JsonWriter jsonWriter = new JsonTextWriter(textWriter))
{
    jsonWriter.StringEscapeHandling = StringEscapeHandling.Default;
    jsonWriter.WriteValue("hello<>!&é");
}
Console.WriteLine(textWriter.ToString()); // "hello<>!&é"

textWriter = new StringWriter();
using (JsonWriter jsonWriter = new JsonTextWriter(textWriter))
{
    jsonWriter.StringEscapeHandling = StringEscapeHandling.EscapeHtml;
    jsonWriter.WriteValue("hello<>!&é");
}
Console.WriteLine(textWriter.ToString()); // "hello\u003c\u003e!\u0026é"

textWriter = new StringWriter();
using (JsonWriter jsonWriter = new JsonTextWriter(textWriter))
{
    jsonWriter.StringEscapeHandling = StringEscapeHandling.EscapeNonAscii;
    jsonWriter.WriteValue("hello<>!&é");
}
Console.WriteLine(textWriter.ToString()); // "hello<>!&\u00e9"

var outputStream = new MemoryStream();
using (var jsonWriter = new Utf8JsonWriter(outputStream))
{
    jsonWriter.WriteStringValue("hello<>!&é");
}
Console.WriteLine(Encoding.UTF8.GetString(outputStream.ToArray())); // "hello\u003C\u003E!\u0026\u00E9"
```

Also, it might be worth noting that the following customizations are not supported with `Utf8JsonWriter`:
[Indentation](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonTextWriter_Indentation.htm) - how many indentation chars to use, we always use 2
[IndentChar](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonTextWriter_IndentChar.htm) - which char to use for indentation, we always use whitespace character
[QuoteChar](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonTextWriter_QuoteChar.htm) - which char to use to surround string values, we always use double quote (otherwise the JSON produced would be invalid)
[QuoteName](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonTextWriter_QuoteName.htm) - whether or not to surround property names with quotes, we always quote them (otherwise the JSON produced would be invalid)

If you rely on any of these features/customization, you should keep using `Newtonsoft.Json` or consider changing your output format to be JSON compliant or one that can be written by `Utf8JsonWriter` (i..e can't customize the white space/indentation behavior).